Refactor

langdonholmes opened the pull request:
Full rewrite of Accordion widget, mostly copied from Drupal's [Bootstrap Accordion plugin](https://git.drupalcode.org/project/ckeditor5_bootstrap_accordion/-/tree/1.x). Icons are borrowed from that project as well.

Currently, the blue caret beside the AccordionItemHeader is non-functional. I think this would require some JavaScript (such as Bootstrap's JS modules), but I'm not sure it's worth the trouble for the moment. I imagine most authors will just leave the full Accordion open in most cases.

![image](https://github.com/user-attachments/assets/4c1a7c3f-0071-47ba-8d0e-1acafaa47ce0)

I spent some time with the CKEditor code on this one, and I think this can serve as a model as we develop future plugins. @theannachen Would be great if you could take some time reviewing this PR. Let me know if you are struggling to understand any aspects of it. I hope that we can discuss this over Slack and also work together to add additional comments to help us understand and re-purpose this code in the future.

Notice also that I left the following comment in the MDX parsing file, inspired by the Bootstrap Accordion plugin. I think we should try to do this for existing and future TurndownService Rules:
```
// Accordions
/* DataModel
  <div class="accordion accordion-items-stay-open" data-accordion-id="">
    <div class="accordion-item">
        <div class="accordion-header">
            <a class="accordion-button" href="#">
                Accordion Title
            </a>
        </div>
        <div class="accordion-collapse collapse show">
            <div class="accordion-body">
                <p>Accordion Content</p>
            </div>
        </div>
    </div>
  </div>
*/
/* MDX Export
  <Accordion value="first" className="prose dark:prose-invert">
    <AccordionItem value="1" title="Accordion Title">
        <p>Accordion Content</p>
    </AccordionItem>
  </Accordion>
*/
```

Minor changes:
- Removed debugging console.log() calls throughout the repository. Console.log() shouldn't be used for normal info messages.
- Removed code that has been commented out in several places throughout the repository. Let's commit to removing code when it is no longer needed. We have git to recover it if needed.

Note that we need to be careful about merging this, as it could potentially break existing Accordions. A given chunk's MDX field will not be affected until its parent page is edited and saved. Vercel deployments would not be affected unless the previous happens and Github Publish is used without repairing the Accordions.

theannachen commented:
Great work on this! Did you use prettier or something similar? I think if we are doing lint stuff, it would be useful to add some documentation in the README so we can keep it consistent between developers.

theannachen commented on src/plugins/auto-content/server/services/mdx-service.js:
```diff
@@ -68,82 +23,102 @@ function stringifyAttributes(element, separator = " ") {
+
+      // Get innerHtml from body. This has not been thoroughly tested, but is intended
+      // to preserve e.g., lists, linebreaks that would be lost with .textContent.
+      const content = item.querySelector(".accordion-body").innerHTML;
```
Currently, the parsing creates AccordionItems with <p> and stuff inside of them instead of just the raw text. The frontend only takes in raw text inside. If you want, I can add this change.

langdonholmes replied:
Sorry, looks like some of this comment is missing.

Are you sure that NextJS only permits plain text? What about text formatting (`<i></i>`)? Lists? Even if NextJS does not allow any text formatting (which would be a problem), using `.textContent` will also destroy linebreaks, so I don't think that is an acceptable solution in any case.

theannachen replied:
Sorry it seems like the comment got messed up when I was posting it. I think this makes sense but certain things like the Info boxes etc inside of them which was more my concern.

langdonholmes replied:
Gotcha, this was resolved by other changes.

theannachen commented on src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Accordion/AccordionCommands.js:
```diff
@@ -0,0 +1,261 @@
+    const accordionItem = this.accordionItem;
+    const value = options ? options.value : "after";
+    model.change((writer) => {
+      const newAccordionItem = createAccordionItem(
```
The accordion items allow everything in them (including other custom blocks). This should be prevented.

langdonholmes replied:
Yes, we should come up with a reusable schema to prevent nesting of custom elements, as this is a reoccurring problem across our plugins.

I had trouble preventing nested accordions using schemas, so I hacked a fix here, but we should probably figure out how to get it working with schemas so we don't have to manually write checks for all components.

langdonholmes replied:
I'll create an issue for this because I think this is a larger concern that applies to multiple plugins. Do you have any ideas?

For one of the recent plugins, we created a naked textblock (rather than a `$root` element with a `<p>` child), which does prevent nesting, but also prevents elements that we might like to allow, such as lists.

theannachen replied:
I believe that the lists were causing problems in the frontend when displaying. I'll contact Qiushi about what exactly should be and shouldn't be allowed in these again.

langdonholmes commented:
> Great work on this! Did you use prettier or something similar? I think if we are doing lint stuff, it would be useful to add some documentation in the README so we can keep it consistent between developers.

Great point. Slightly off-topic for this PR, but I added Prettier and ESLint configs and ran them on the entire repo. These are very much a work-in-progress, and I feel that it will be difficult to maintain a high degree of consistency in this repo because we have code coming from multiple places (Strapi TSX, our JS code, CKEditor DLL shenanigans). Also updated the readme.

To get it working, I had to set many linter rules to "warn" instead of "error", which partially defeats the point of using a linter, but that's just the state of the repo at the moment, and I think something is better than nothing. We can try to improve on this over time.

langdonholmes commented:
Lots of tweaks here. I mainly changed the way we are handling schemas in our plugins. This was largely accomplished by specifying `allowedChildren` rather than naively inheriting a complex schema like `$root`. As a benefit, this allowed me to remove some overly complex schema checks, which are intended for more advanced scenarios that can't be covered with simple schema definitions.

I also changed the MDX parsing to allow for lists on a few elements, but please note that this changes the parsing behavior! I do not understand why we were adding `<br/>`, so please review the new outputs and let me know if this is needed. Example for a Callout with two paragraphs:

Old Output (HTML + Markdown):
```
<Callout>
Paragraph 1 <br/>
Paragraph 2
</Callout>
```

New output (pure Markdown):
```
<Callout>
Paragraph 1

Paragraph 2
</Callout>
```

langdonholmes commented on src/plugins/ckeditor/admin/src/components/CKEditorInput/plugins/Info/InfoEditing.js:
```diff
@@ -77,7 +63,7 @@ export default class InfoEditing extends Plugin {
       view: (modelElement, { writer: viewWriter }) => {
         const section = viewWriter.createContainerElement("section", {
-          class: "Callout",
+          class: "Callout", // TODO: Fix the class name and CSS styles!
```
@theannachen Can you work on this? A commit to this PR would be fine.

langdonholmes commented on src/plugins/auto-content/server/services/mdx-service.js:
```diff
@@ -189,33 +162,32 @@ turndownService.addRule("image", {
-turndownService.addRule('convertLineBreaks', {
-  filter: 'br',
+// Converts linebreaks
+// TODO: Explain why this is needed.
```
@theannachen Still need help understanding this.

theannachen replied:
The parser I believe transforms them to <br> but since jsx components are required to all be self-closing, they need to be transformed into a <br/>

langdonholmes replied:
Broken comment?

The parser by default replaces 'br' with `  \n` (two spaces and a linebreak).

https://github.com/mixmark-io/turndown/blob/cc73387fb707e5fb5e1083e94078d08f38f3abc8/src/commonmark-rules.js#L13

langdonholmes replied:
@theannachen Just a reminder that I need some closure on this ;-)
